content/post/performance-metrics-overview.md

```diff
@@ -0,0 +1,111 @@
+
+1. High-Throughput
+2. Low-Latency
+3. 1/3 of the infrastructure costs/needs compared to existing solutions.
+
+Proving to our users that we were delivering on these principles was an essential part of the early development cycle of Wallaroo. How could we prove we were delivering on these principles? We needed to capture metrics. Metrics are an excellent way to spot bottlenecks in your system. Providing meaningful metrics to our users would mean they could make quicker iterations in their development cycle by being able to spot specific bottlenecks in various locations of our system. An added benefit that came from capturing these metrics was that it also sped up the Wallaroo development cycle as we were able to use them as guidelines to [measure the “cost” of every feature](https://blog.wallaroolabs.com/2017/06/whats-the-secret-sauce/#measure-the-cost-of-every-feature).
+
+In the end, what came of our metrics capturing process became a web UI that provides Wallaroo users introspection for various parts of the applications they develop. Here's what the result looks like:
+
+![application dashboard page](/images/post/performance-metrics-overview/application-dashboard-page.png)
+
+In this post, I'll cover some of the design decisions we made regarding our metrics capturing system to maintain the high performance our system was promising.
+
+## Metrics Aren't Free
+
+Adding introspection via metrics to your system inherently adds an overhead. However, the impact of that overhead is determined by how you capture metrics and what information you choose to capture. Not only is there a ton of information to capture, but there are also many different ways to capture it. The biggest problem we face while determining how we capture metrics is maintaining a low overhead. A computationally heavy metrics capturing system could mean lower throughput, higher latencies, and costlier infrastructure, all things we want to avoid. When approaching this type of problem, you need to determine how much information you need to capture to provide meaningful details via your metrics while minimizing the performance impact on your system.
+
+## Information We Want to Convey with our Metrics
+
+A Wallaroo application broken down to its simplest form is composed of the following components:
+
+- **Computations:** code that transforms an input of some type to an output of some type. In Wallaroo, there are stateless computations and what we call “state computations” that also operate on in-memory state.
+- **Pipelines:** a sequence of computations and/or state computations originating from a source and optionally terminating in a sink.
+
+These components can be on one or more Wallaroo **workers** (process). By capturing metrics for these components, we give Wallaroo users a granular look into various parts of the applications they develop.
+
+The information we want to convey to our users with the metrics we capture are the following:
```

Comment:
> The information we want to convey to our users with the metrics we capture are the following:

should be

> The information we want to convey to our users with the metrics we capture **is** the following:

because the `is` is referring to `the information` and not to `the metrics`.